src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/WireMessage.java

```diff
@@ -120,37 +69,127 @@ public static WireMessage create(ConnectionHandler connection, DataInputStream i
+   */
+  protected abstract void sendPayload() throws Exception;
+
+  /**
+   * Used for logging.
+   *
+   * @return The official name of the wire message.
+   */
+  protected abstract String getMessageName();
+
+  /**
+   * Used for logging.
+   *
+   * @return Payload metadata.
+   */
+  protected abstract String getPayloadString();
+
+  /**
+   * Used for logging.
+   *
+   * @return Message Identifier (int for Bootstrap, char otherwise).
+   */
+  protected abstract String getIdentifier();
+
+  @Override
+  public String toString() {
+    return new MessageFormat("> Received Message: ({0}) {1}, with Payload: '{'{2}'}'")
+        .format(new Object[]{
+            this.getIdentifier(),
+            this.getMessageName(),
+            this.getPayloadString()
+        });
+  }
 
   /**
-   * Read the remainder of the received message. To be called once metdata for the message is read.
+   * Read the remainder of the received message. To be called once metadata for the message is read.
    * Metadata is designated as type, length, etc.
    *
-   * @param input The input stream containing the remainder of the message.
    * @return The remainder of the message in String format.
    * @throws Exception If reading fails in any way.
    */
-  protected String read(DataInputStream input) throws Exception {
-    return PGWireProtocol.readString(input, this.length - this.remainder);
+  protected String readAll() throws Exception {
+    return read(this.length - this.getHeaderLength());
   }
 
+  /**
+   * Reads a fixed-length string from teh saved stream. The string still needs to be
+   * null-terminated, but the read is more efficient, as we can read the entire string in one go
+   * instead of continuously checking for a null-terminator.
+   *
+   * @param length The number of bytes to read.
+   * @return the string.
+   * @throws IOException if an error occurs while reading from the stream, or if no null-terminator
+   * is found at the end of the string.
+   */
+  public String read(int length) throws IOException {
+    byte[] buffer = new byte[length - 1];
+    this.inputStream.readFully(buffer);
+    byte zero = inputStream.readByte();
+    if (zero != (byte) 0) {
+      throw new IOException("String was not null-terminated");
+    }
+    return new String(buffer, StandardCharsets.UTF_8);
+  }
 
-  public enum PreparedType {
-    Portal,
-    Statement;
-
-    static PreparedType prepareType(char type) {
-      switch (type) {
-        case ('P'):
-          return PreparedType.Portal;
-        case ('S'):
-          return PreparedType.Statement;
-        default:
-          throw new IllegalArgumentException("Unknown Statement type!");
+  /**
+   * Reads a null-terminated string from a {@link DataInputStream}.
+   *
+   * @return the string.
+   * @throws IOException if an error occurs while reading from the stream, or if no null-terminator
+   * is found before the end of the stream.
+   */
+  public String readString() throws IOException {
```

Comment:
feels weird to be implementing something this basic like reading a null-terminated string off the wire.  I'm not as up on my JDK as I used to be - nothing already written here?
is 128 bytes the right starting point?  Is it better to read as much as available and put back the bytes after the null byte?

Reply:
Technically this was inherited from the previous pgadapter and used equally in the JDBC driver, but I couldn't find anything else that was embedded easily - that is to say, this was the cleanest and most readable way to do it. I did do a quick profile of that, and most messages aren't much longer than 128, so I think it's an okay one to start doubling.